src/torchmetrics/functional/classification/stat_scores.py

```diff
@@ -16,8 +16,453 @@
+        if ignore_index is not None and not ignore_in:
+            preds = preds.clone()
+            target = target.clone()
+            idx = target == ignore_index
+            preds[idx] = num_classes
+            target[idx] = num_classes
+
+        if top_k > 1:
+            preds_oh = select_topk(preds, topk=top_k, dim=1).movedim(1, -1)
+        else:
+            preds_oh = torch.nn.functional.one_hot(
+                preds, num_classes + 1 if ignore_index is not None and not ignore_in else num_classes
+            )
+        target_oh = torch.nn.functional.one_hot(
+            target, num_classes + 1 if ignore_index is not None and not ignore_in else num_classes
+        )
+        if ignore_index is not None:
+            if 0 <= ignore_index <= num_classes - 1:
+                target_oh[target == ignore_index, :] = -1
+            else:
+                preds_oh = preds_oh[..., :-1]
+                target_oh = target_oh[..., :-1]
+                target_oh[target == num_classes, :] = -1
+        sum_dim = [0, 1] if multidim_average == "global" else [1]
+        tp = ((target_oh == preds_oh) & (target_oh == 1)).sum(sum_dim)
+        fn = ((target_oh != preds_oh) & (target_oh == 1)).sum(sum_dim)
+        fp = ((target_oh != preds_oh) & (target_oh == 0)).sum(sum_dim)
+        tn = ((target_oh == preds_oh) & (target_oh == 0)).sum(sum_dim)
+        return tp, fp, tn, fn
+    else:
+        preds = preds.flatten()
+        target = target.flatten()
+        if ignore_index is not None:
+            idx = target != ignore_index
+            preds = preds[idx]
+            target = target[idx]
+        unique_mapping = (target * num_classes + preds).to(torch.long)
+        bins = _bincount(unique_mapping, minlength=num_classes**2)
+        confmat = bins.reshape(num_classes, num_classes)
+        tp = confmat.diag()
+        fp = confmat.sum(0) - tp
+        fn = confmat.sum(1) - tp
+        tn = confmat.sum() - (fp + fn + tp)
+        return tp, fp, tn, fn
+
+
+def _multiclass_stat_scores_compute(
+    tp: Tensor, fp: Tensor, tn: Tensor, fn: Tensor, average: str = "micro", multidim_average: str = "global"
+) -> Tensor:
+
+    res = torch.stack([tp, fp, tn, fn, tp + fn], dim=-1)
+    sum_dim = 0 if multidim_average == "global" else 1
+    if average == "micro":
+        return res.sum(sum_dim)
+    elif average == "macro":
+        return res.float().mean(sum_dim)
+    elif average == "weighted":
+        weight = tp + fn
+        if multidim_average == "global":
+            return (res * (weight / weight.sum()).reshape(*weight.shape, 1)).sum(sum_dim)
+        else:
+            return (res * (weight / weight.sum(-1, keepdim=True)).reshape(*weight.shape, 1)).sum(sum_dim)
+    elif average is None or average == "none":
+        return res
+
+
+def multiclass_stat_scores(
+    preds: Tensor,
+    target: Tensor,
+    num_classes: int,
+    average: str = "micro",
+    top_k: int = 1,
+    multidim_average: str = "global",
```

Comment:
Verbosity for public API
```suggestion
    average: Optional[Literal["micro", "macro", "weighted", "none"]] = "micro",
    top_k: int = 1,
    multidim_average: Literal["global", "samplewise"] = "global",
```